diff --git a/tensorflow/core/distributed_runtime/base_rendezvous_mgr.cc b/tensorflow/core/distributed_runtime/base_rendezvous_mgr.cc
index de6e4b4a7c5..3e5a4bf1888 100644
--- a/tensorflow/core/distributed_runtime/base_rendezvous_mgr.cc
+++ b/tensorflow/core/distributed_runtime/base_rendezvous_mgr.cc
@@ -102,6 +102,20 @@ Status BaseRendezvousMgr::RecvLocal(int64 step_id,
 }
 
 void BaseRendezvousMgr::Cleanup(int64 step_id) {
+  // Delay the cleanup by 1024 steps for ByteScheduler
+  int64 step_to_clean;
+  {
+    mutex_lock l(mu_);
+    if (steps_to_clean_.size() >= 1024) {
+      step_to_clean = steps_to_clean_.front();
+      steps_to_clean_.pop_front();
+      steps_to_clean_.push_back(step_id);
+    } else {
+      steps_to_clean_.push_back(step_id);
+    }
+  }
+  step_id = step_to_clean;
+
   Rendezvous* rendez = nullptr;
   {
     mutex_lock l(mu_);
diff --git a/tensorflow/core/distributed_runtime/base_rendezvous_mgr.h b/tensorflow/core/distributed_runtime/base_rendezvous_mgr.h
index c5a56f324ab..e0656a49fd7 100644
--- a/tensorflow/core/distributed_runtime/base_rendezvous_mgr.h
+++ b/tensorflow/core/distributed_runtime/base_rendezvous_mgr.h
@@ -16,6 +16,7 @@ limitations under the License.
 #ifndef TENSORFLOW_CORE_DISTRIBUTED_RUNTIME_BASE_RENDEZVOUS_MGR_H_
 #define TENSORFLOW_CORE_DISTRIBUTED_RUNTIME_BASE_RENDEZVOUS_MGR_H_
 
+#include <deque>
 #include <string>
 #include <unordered_set>
 
@@ -102,6 +103,7 @@ class BaseRendezvousMgr : public RendezvousMgrInterface {
   // Not owned.
   const WorkerEnv* const worker_env_;
 
+  std::deque<int64> steps_to_clean_ GUARDED_BY(mu_);
   mutex mu_;
   Table table_ GUARDED_BY(mu_);
 
diff --git a/tensorflow/workspace.bzl b/tensorflow/workspace.bzl
index 527d2a175dc..a9ca4f351a6 100755
--- a/tensorflow/workspace.bzl
+++ b/tensorflow/workspace.bzl
@@ -462,6 +462,7 @@ def tf_workspace(path_prefix = "", tf_repo_name = ""):
     # WARNING: make sure ncteisen@ and vpai@ are cc-ed on any CL to change the below rule
     tf_http_archive(
         name = "grpc",
+       patch_file = clean_dep("//third_party:Rename-gettid-functions.patch"),
         sha256 = "1aa84387232dda273ea8fdfe722622084f72c16f7b84bfc519ac7759b71cdc91",
         strip_prefix = "grpc-69b6c047bc767b4d80e7af4d00ccb7c45b683dae",
         system_build_file = clean_dep("//third_party/systemlibs:grpc.BUILD"),
diff --git a/third_party/icu/workspace.bzl b/third_party/icu/workspace.bzl
index f100836b410..74ed39efe93 100644
--- a/third_party/icu/workspace.bzl
+++ b/third_party/icu/workspace.bzl
@@ -11,7 +11,7 @@ def repo():
     third_party_http_archive(
         name = "icu",
         strip_prefix = "icu-release-62-1",
-        sha256 = "e15ffd84606323cbad5515bf9ecdf8061cc3bf80fb883b9e6aa162e485aa9761",
+        sha256 = "86b85fbf1b251d7a658de86ce5a0c8f34151027cc60b01e1b76f167379acf181",
         urls = [
             "https://mirror.bazel.build/github.com/unicode-org/icu/archive/release-62-1.tar.gz",
             "https://github.com/unicode-org/icu/archive/release-62-1.tar.gz",